lib/SimplifyPointerBitcastPass.cpp

```diff
@@ -196,11 +196,15 @@ bool SimplifyPointerBitcastPass::runOnBitcastFromBitcast(Module &M) const {
     for (BasicBlock &BB : F) {
       for (Instruction &I : BB) {
         // If we have a bitcast instruction...
-        if (auto Bitcast = dyn_cast<BitCastInst>(&I)) {
+        if (auto *Bitcast = dyn_cast<BitCastInst>(&I)) {
           // ... whose source is a bitcast instruction...
-          if (isa<BitCastInst>(Bitcast->getOperand(0))) {
-            // ... record the bitcast as something we need to process.
-            WorkList.push_back(Bitcast);
+          if (auto *OtherBitcast = dyn_cast<BitCastInst>(Bitcast->getOperand(0))) {
+            // ... record the bitcast as something we need to process...
+            if (OtherBitcast->getType() == Bitcast->getType() ||
```

Comment:
This seems overly conservative.  The real problem seems to be the way the cast instruction is created. It should instead use `CastInst::Create(Instruction::BitCast,...)`.

Reply:
That works, indeed. I wonder though, should the pass be renamed to, say, `SimplifyBitcastPass`? The current name is the main reason my patch ignores non-trivial, non-pointer casts.

Reply:
It had already grown past that point with the trivial bitcast transformation. I'm ok to leave as is or change it in this PR.